src/v/storage/mvlog/batch_collecting_stream_utils.h

```diff
@@ -0,0 +1,42 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+
+#include "storage/mvlog/entry_stream.h"
+#include "storage/mvlog/reader_outcome.h"
+
+namespace storage::experimental::mvlog {
+
+class batch_collector;
+class entry_stream;
+
+enum class collect_stream_outcome {
+    // The batch collector is full.
+    buffer_full,
+
+    // There are no more entries in the stream.
+    end_of_stream,
+
+    // The batch collection is complete, e.g. because we have reached the
+    // desired offset.
+    stop,
+};
+std::ostream& operator<<(std::ostream&, collect_stream_outcome);
+
+// Parses and collects the record batches from the given entry stream.
+// Ignores other kinds of entries.
+ss::future<result<collect_stream_outcome, errc>>
+collect_batches_from_stream(entry_stream* entries, batch_collector* collector);
```

Comment:
I noticed a few places where this uses raw pointers instead of references or managed pointers. Is there a reason for that?

Reply:
I generally prefer raw pointers for in-out and out parameters, a la Google style guide.

As for member pointer variables, unless there's some complicated lifetime relationship between classes, I prefer having a single owner being denoted by a unique_ptr and passing around raw pointers. I'm hoping for the lifetimes of these obejcts to be straightforward enough that it's easily verifiable (e.g. readable_segment is long lived, segment_readers must not outlive the readable_segment, etc), but if that ends up not being the case, I wouldn't be against some kind of smart pointer

Reply:
@andrwng I see google style guide recommending references https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs
>non-optional output and input/output parameters should usually be references (which cannot be null)

Reply:
Well I'll be! Let me fix